src/Compilers/Core/Portable/Collections/UnionCollection.cs

```diff
@@ -68,6 +85,38 @@ private UnionCollection(ImmutableArray<ICollection<T>> collections)
             _collections = collections;
         }
 
+        private static ImmutableArray<ICollection<T>> Flatten(ImmutableArray<ICollection<T>> collections)
+        {
+            if (!hasUnionCollection(collections))
+                return collections;
+
+            var builder = ArrayBuilder<ICollection<T>>.GetInstance();
+            foreach (var collection in collections)
+            {
+                if (collection is UnionCollection<T> unionCollection)
+                {
+                    builder.AddRange(unionCollection._collections);
+                }
+                else
+                {
+                    builder.Add(collection);
```

Comment:
potentially ignore empty collections (like what is done above)?

Reply:
At first I was thinking these are read-only but not necessarily immutable, but with the code above it does seem like a reasonable optimization.

Reply:
actually, my guess is that it's not.  in practice these will never be empty.  :)